File: competitive_programming/leetcode/03_cpp/7_reverse_integer.cpp
#include <algorithm>
#include <cassert>
#include <cmath>
#include <cstdint>
#include <iostream>
#include <set>
#include <string>
#include <vector>

int reverseInteger(int x) {
    int sign = x < 0 ? -1 : 1;
    long long value = std::abs(static_cast<long long>(x));
    long long reversed = 0;
    while (value > 0) { reversed = reversed * 10 + value % 10; value /= 10; }
    reversed *= sign;
    return (reversed < -(1LL << 31) || reversed > (1LL << 31) - 1) ? 0 : static_cast<int>(reversed);
}

int main() {
    assert(reverseInteger(123) == 321 && reverseInteger(-123) == -321 && reverseInteger(120) == 21);
    assert(reverseInteger(1534236469) == 0);
    std::cout << "7 reverse integer ok\n";
    return 0;
}

File: competitive_programming/leetcode/09_rust/7_reverse_integer.rs
fn reverse_integer(x: i32) -> i32 {
    let sign: i64 = if x < 0 { -1 } else { 1 };
    let mut value = (x as i64).abs();
    let mut reversed: i64 = 0;
    while value > 0 { reversed = reversed * 10 + value % 10; value /= 10; }
    reversed *= sign;
    if reversed < -(1i64 << 31) || reversed > (1i64 << 31) - 1 { 0 } else { reversed as i32 }
}

fn main() {
    assert_eq!(reverse_integer(123), 321);
    assert_eq!(reverse_integer(-123), -321);
    assert_eq!(reverse_integer(120), 21);
    assert_eq!(reverse_integer(1534236469), 0);
    println!("7 reverse integer ok");
}
